src/languages/nix.js

```diff
@@ -6,88 +6,372 @@ Website: http://nixos.org/nix
+  const OPERATOR_WITHOUT_MINUS_REGEX = regex.either(...[
+    '==',
+    '=',
+    '\\+\\+',
+    '\\+',
+    '<=',
+    '<\\|',
+    '<',
+    '>=',
+    '>',
+    '->',
+    '//',
+    '/',
+    '!=',
+    '!',
+    '\\|\\|',
+    '\\|>',
+    '\\?',
+    '\\*',
+    '&&',
+  ]);
+
+  const OPERATOR = {
+    scope: 'operator',
+    match: regex.concat(OPERATOR_WITHOUT_MINUS_REGEX, /(?!-)/),
+    relevance: 0,
   };
+
+  // '-' is being handled by itself to ensure we are able to tell the difference
+  // between a dash in an identifier and a minus operator
+  const NUMBER = {
+    scope: 'number',
+    match: new RegExp(`${hljs.NUMBER_RE}(?!-)`),
+    relevance: 0,
+  };
+  const MINUS_OPERATOR = {
+    variants: [
+      {
+        scope: 'operator',
+        beforeMatch: /\s/,
+        // The (?!>) is used to ensure this doesn't collide with the '->' operator
+        begin: /-(?!>)/,
+      },
+      {
+        begin: [
+          new RegExp(`${hljs.NUMBER_RE}`),
+          /-/,
+          /(?!>)/,
+        ],
+        beginScope: {
+          1: 'number',
+          2: 'operator'
+        },
+      },
+      {
+        begin: [
+          OPERATOR_WITHOUT_MINUS_REGEX,
+          /-/,
+          /(?!>)/,
+        ],
+        beginScope: {
+          1: 'operator',
+          2: 'operator'
+        },
+      },
+    ],
+    relevance: 0,
+  };
+
   const ATTRS = {
-    begin: /[a-zA-Z0-9-_]+(\s*=)/,
+    beforeMatch: /(^|\{|;)\s*/,
+    begin: new RegExp(`${IDENTIFIER_REGEX}(\\.${IDENTIFIER_REGEX})*\\s*=(?!=)`),
     returnBegin: true,
     relevance: 0,
     contains: [
       {
-        className: 'attr',
-        begin: /\S+/,
-        relevance: 0.2
+        scope: 'attr',
+        match: new RegExp(`${IDENTIFIER_REGEX}(\\.${IDENTIFIER_REGEX})*(?=\\s*=)`),
+        relevance: 0.2,
       }
-    ]
+    ],
+  };
+
+  const NORMAL_ESCAPED_DOLLAR = {
+    scope: 'char.escape',
+    match: /\\\$/,
+  };
+  const INDENTED_ESCAPED_DOLLAR = {
+    scope: 'char.escape',
+    match: /''\$/,
+  };
+  const ANTIQUOTE = {
+    scope: 'subst',
+    begin: /\$\{/,
+    end: /\}/,
+    keywords: KEYWORDS,
+  };
+  const ESCAPED_DOUBLEQUOTE = {
+    scope: 'char.escape',
+    match: /'''/,
+  };
+  const ESCAPED_LITERAL = {
+    scope: 'char.escape',
+    match: /\\(?!\$)./,
   };
   const STRING = {
-    className: 'string',
-    contains: [ ESCAPED_DOLLAR, ANTIQUOTE ],
+    scope: 'string',
     variants: [
       {
         begin: "''",
-        end: "''"
+        end: "''",
+        contains: [
+          INDENTED_ESCAPED_DOLLAR,
+          ANTIQUOTE,
+          ESCAPED_DOUBLEQUOTE,
+          ESCAPED_LITERAL,
+        ],
       },
       {
         begin: '"',
-        end: '"'
-      }
-    ]
+        end: '"',
+        contains: [
+          NORMAL_ESCAPED_DOLLAR,
+          ANTIQUOTE,
+          ESCAPED_LITERAL,
+        ],
+      },
+    ],
+  };
+
+  const URL = {
+    scope: 'link',
+    match: /https?:\/\/[A-Za-z0-9\-._~:/?#[\]@!$&'()*+,=]+/,
```

Comment:
Does nix support ftp, ssh, or other protocols?

Reply:
Nix actually supports the entirety of rfc2396 (see last paragraph of https://nix.dev/manual/nix/2.24/language/syntax#string-literal), but I wasn't able to make direct use of the regex provided in the rfc because it kept picking up all sorts of false positives.

The feature is quite discouraged these days, mostly replaced with strings, and with no instances left that I'm aware of in nixpkgs (the largest and most important repository for the ecosystem). So I used a slightly suboptimal regex, hoping to catch some basic instances.

I could add a few more common protocols, but I'm not entirely sure where to stop.

Reply:
If it's deprecated/discouraged perhaps lets drop it altogether so people don't complain it's not working in their own special use cases... What are they doing now instead, quoted strings?

Reply:
I have removed the commits related to it. It will pick instances up like a literal path, but maybe it's better to actively not support it (at least for time being, until someone comes up with a better regex). Quoted strings is the recommended replacement, yes.

``` nix
Nix 2.24.9
Type :? for help.
nix-repl> https://nixos.org == "https://nixos.org"
true
```

Reply:
I think this is more than OK, as there are linters that will automatically fix it and both active lsp implementation warn and provide code actions for fixes:

![image](https://github.com/user-attachments/assets/7fae51d5-7b96-47ed-ba7e-71c23e56c046)